packages/nitro-protocol/contracts/NitroAdjudicator.sol

```diff
@@ -110,34 +40,64 @@ contract NitroAdjudicator is IAdjudicator, ForceMove {
+     * @notice Liquidates all assets for the channel
+     * @dev Liquidates all assets for the channel
      * @param channelId Unique identifier for a state channel
      * @param outcomeBytes abi.encode of an array of Outcome.OutcomeItem structs.
+     * @param stateHash stored state hash for the channel
+     * @param challengerAddress stored challenger address for the channel
      */
-    function _transferAllFromAllAssetHolders(bytes32 channelId, bytes memory outcomeBytes)
-        internal
-    {
+    function transferAllAssets(
+        bytes32 channelId,
+        bytes memory outcomeBytes,
+        bytes32 stateHash,
+        address challengerAddress
+    ) public {
+        // checks
+        _requireChannelFinalized(channelId);
+        _requireMatchingFingerprint(
+            stateHash,
+            challengerAddress,
+            keccak256(outcomeBytes),
+            channelId
+        );
         Outcome.OutcomeItem[] memory outcome = abi.decode(outcomeBytes, (Outcome.OutcomeItem[]));
 
-        for (uint256 i = 0; i < outcome.length; i++) {
+        for (uint256 assetIndex = 0; assetIndex < outcome.length; assetIndex++) {
             Outcome.AssetOutcome memory assetOutcome = abi.decode(
-                outcome[i].assetOutcomeBytes,
+                outcome[assetIndex].assetOutcomeBytes,
                 (Outcome.AssetOutcome)
             );
-            if (assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Allocation) {
-                AssetHolder(outcome[i].assetHolderAddress).transferAllAdjudicatorOnly(
-                    channelId,
-                    assetOutcome.allocationOrGuaranteeBytes
-                );
-            } else {
-                revert('AssetOutcome not an allocation');
-            }
+            require(
+                assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Allocation,
+                '!allocation'
+            );
+            Outcome.AllocationItem[] memory allocation = abi.decode(
+                assetOutcome.allocationOrGuaranteeBytes,
+                (Outcome.AllocationItem[])
+            );
+            address asset = outcome[assetIndex].asset;
+            uint256 initialHoldings;
+            // update allocation in place, to the new allocation returned by _transfer
+            (allocation, initialHoldings) = _transfer(
+                asset,
+                channelId,
+                allocation,
+                new uint256[](0)
+            );
+            outcome[assetIndex].assetOutcomeBytes = abi.encode(
+                Outcome.AssetOutcome(Outcome.AssetOutcomeType.Allocation, abi.encode(allocation))
+            );
+            emit AllocationUpdated(channelId, assetIndex, initialHoldings);
         }
+        outcomeBytes = abi.encode(outcome);
+        bytes32 outcomeHash = keccak256(outcomeBytes);
+        _updateFingerprint(channelId, stateHash, challengerAddress, outcomeHash);
```

Comment:
I think probably yes. We have some discussion around this already https://github.com/statechannels/statechannels/pull/3394 . There are security considerations involved, and note also that clearing slots is not as beneficial as it once was (post London). 

I would rather not make the change now, because it isn't necessary as a part of this already huge PR and muddies the gas comparison with the base branch (which currently does not have that optimization).